packages/engine/paima-funnel/src/cde/cardanoProjectedNFT.ts

```diff
@@ -0,0 +1,67 @@
+import type {
+  CdeCardanoProjectedNFTDatum,
+  ChainDataExtensionCardanoProjectedNFT,
+  ChainDataExtensionDatum,
+} from '@paima/sm';
+import { ChainDataExtensionDatumType, DEFAULT_FUNNEL_TIMEOUT, timeout } from '@paima/utils';
+import { Routes, query } from '@dcspark/carp-client/client/src';
+import type { ProjectedNftRangeResponse } from '@dcspark/carp-client/shared/models/ProjectedNftRange';
+
+export default async function getCdeProjectedNFTData(
+  url: string,
+  extension: ChainDataExtensionCardanoProjectedNFT,
+  fromAbsoluteSlot: number,
+  toAbsoluteSlot: number,
+  getBlockNumber: (slot: number) => number
+): Promise<ChainDataExtensionDatum[]> {
+  const events = await timeout(
+    query(url, Routes.projectedNftEventsRange, {
+      range: { minSlot: fromAbsoluteSlot, maxSlot: toAbsoluteSlot },
+    }),
+    DEFAULT_FUNNEL_TIMEOUT
+  );
+
+  return events
+    .map(e => eventToCdeDatum(e, extension, getBlockNumber(e.actionSlot)))
+    .filter(e => e != null)
+    .map(e => e!);
+}
+
+function eventToCdeDatum(
+  event: ProjectedNftRangeResponse[0],
+  extension: ChainDataExtensionCardanoProjectedNFT,
+  blockNumber: number
+): CdeCardanoProjectedNFTDatum | null {
+  if (
+    event.actionTxId === null ||
+    event.actionTxId == '' ||
+    event.status === null ||
+    event.status == ''
+  ) {
+    return null;
+  }
+
+  return {
+    cdeId: extension.cdeId,
+    cdeDatumType: ChainDataExtensionDatumType.CardanoProjectedNFT,
+    blockNumber,
+    payload: {
+      ownerAddress: event.ownerAddress != null ? event.ownerAddress : '',
+
+      actionTxId: event.actionTxId,
```

Comment:
Note that games only care about NFTs getting locked and queued to unlock. We don't care if the user ever actually ends up withdrawing their NFT from the lock later.